### PR TITLE
Solid-Liquid Separator Base Model

### LINKED ISSUE #1218: models.unit_models.MSContactor missing performance results
This breaks serialization.

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
This PR adds a new unit model base class for separating solids and liquids such as filters and thickeners. The unit has two inlets (solid and liquid) and three outlets (solids, retained_liquid (with solids) and recovered liquid).

Docs to come, along with a derived model for thickeners.

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by andrewlee94** [comment]
@adam-a-a You might want to be aware of this for potential use in WaterTAP.

File: docs/reference_guides/model_libraries/generic/unit_models/solid_liquid/index.rst [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Solid-Liquid Unit Models
+========================
+
+This library contains models specific for solid-liquid processing operations (although some of them may also be applicable in other circumstances). Models in this library assume streams consist of separate solid and liquid phases (with separate property packages) and thus all stream are represented by separate solid and liquid ports. Separation of solids and liquids is generally assume to result in a concentrated solid-liquid stream and a clarified liquid stream.
```
Comment (agarciadiego) [inline docs/reference_guides/model_libraries/generic/unit_models/solid_liquid/index.rst:4]: "all streams"

Comment (agarciadiego) [reply docs/reference_guides/model_libraries/generic/unit_models/solid_liquid/index.rst:4]: generally assumed

File: idaes/models/unit_models/solid_liquid/tests/test_sl_separator.py [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+    def test_units(self, model):
+        assert_units_consistent(model)
+
+    @pytest.mark.unit
+    def test_dof(self, model):
+        assert degrees_of_freedom(model) == 0
+
+    @pytest.mark.component
+    def test_structural_issues(self, model):
+        dt = DiagnosticsToolbox(model)
```
Comment (agarciadiego) [inline idaes/models/unit_models/solid_liquid/tests/test_sl_separator.py:179]: Is this going to be standard procedure for tests now? I like it; I'm just asking.

File: idaes/models/unit_models/solid_liquid/tests/test_sl_separator.py [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+from idaes.core import (
+    FlowsheetBlock,
+    MaterialBalanceType,
+    EnergyBalanceType,
+    MomentumBalanceType,
+)
+from idaes.models.unit_models.solid_liquid import SLSeparator
+from idaes.models.unit_models.separator import (
+    Separator,
```
Comment (agarciadiego) [inline idaes/models/unit_models/solid_liquid/tests/test_sl_separator.py:32]: Are all of this imports required?

**Review by agarciadiego — APPROVED** [review]
Just a couple of typos and a question. Otherwise LGTM.

File: docs/reference_guides/model_libraries/generic/unit_models/solid_liquid/sl_separator.rst [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+---------------
+
+The model consists of two parts. The solid phase is represented by a single StateBlock as all solids in the inlet are assumed to report to the concentrated outlet. The liquid phase is represented by a standard :ref:`Separator <reference_guides/model_libraries/generic/unit_models/separator:Separator>` block which splits the incoming liquid between the concentrated and clarified outlets.
+
+The ``SLSeparator`` unit model has a total of 5 ports:
+
+* ``solid_inlet`` representing the solids in the incoming stream,
+* ``liquid_inlet`` representing the liquid in the incoming stream,
+* ``solid_outlet`` representing the solids in the concentrated stream,
+* ``retained_liquid_outlet`` representing the liquid in the concentrated stream, and,
```
Comment (MarcusHolly) [inline docs/reference_guides/model_libraries/generic/unit_models/solid_liquid/sl_separator.rst:21]:
```suggestion
* ``retained_liquid_outlet`` representing the liquid in the concentrated stream
```
Don't think the "and" is necessary in this list. People might assume that some text is supposed to follow but is just missing

**Review by MarcusHolly — APPROVED** [review]
Made a few superficial comments, but it LGTM

**Comment by andrewlee94** [comment]
@MarcusHolly @agarciadiego Would you be able to take a quick look at this again? I added a basic thickener model which derives from the SLSeparator.

File: docs/reference_guides/model_libraries/generic/unit_models/solid_liquid/thickener0d.rst [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+.. math:: max \left( \frac{(Y - Y_{under})}{u(Y)} \right)
+
+where :math:`Y` is the mass based liquid-to-solid ratio, :math:`u(Y)` is the settling velocity of the suspension as a function of :math:`Y` and :math:`Y_{under}` is the liquid-solid ratio at the thickener underflow. Correlations exist which can predict :math:`u(Y)` however in many cases it is necessary to measure this experimentally. Additionally, whilst there are techniques for embedding the maximization operation within an equation-oriented model, these approaches tend to be highly non-linear and require the user to provide good values for scaling parameters and thus have not been implemented yet.
+
+In the current implementation of the model, :math:`Y_{pinch}` and :math:`u_{pinch}` are created as user-defined input variables which should generally be fixed at appropriate values. Users may choose to add a correlation for :math:`u(Y)` if they choose.
+
+Degrees of Freedom
+------------------
+
+The ``Thickener0D`` model has 5 degrees of freedom, which is are generally chosen from:
```
Comment (MarcusHolly) [inline docs/reference_guides/model_libraries/generic/unit_models/solid_liquid/thickener0d.rst:22]:
```suggestion
The ``Thickener0D`` model has 5 degrees of freedom, which are generally chosen from:
```

**Comment by MarcusHolly** [comment]
> @MarcusHolly @agarciadiego Would you be able to take a quick look at this again? I added a basic thickener model which derives from the SLSeparator.

LGTM

**Comment by agarciadiego** [comment]
> > @MarcusHolly @agarciadiego Would you be able to take a quick look at this again? I added a basic thickener model which derives from the SLSeparator.
> 
> LGTM

LGTM

**Pull request merged**